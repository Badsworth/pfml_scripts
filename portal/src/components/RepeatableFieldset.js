import React, { useEffect, useState } from "react";
import Button from "./Button";
import PropTypes from "prop-types";
import RepeatableFieldsetCard from "./RepeatableFieldsetCard";
import uniqueId from "lodash/uniqueId";
import usePreviousValue from "../hooks/usePreviousValue";

/**
 * Used for rendering the same set of content and fields for each
 * item in the "entries" prop.
 */
const RepeatableFieldset = (props) => {
  const { entries } = props;
  const containerRef = React.createRef();
  const entriesAndIds = useEntryIds(entries);
  const previousEntriesLength = usePreviousValue(entriesAndIds.length);

  useEffect(() => {
    if (entriesAndIds.length > previousEntriesLength) {
      // When a new entry is added to the list, focus and scroll it into view.
      const lastEntry = containerRef.current.querySelector(
        ".js-repeated-fieldset-card:last-of-type"
      );
      const focusableElement = lastEntry.querySelector(
        "[tabIndex]:first-child, label"
      );

      if (focusableElement) focusableElement.focus();
    }
  }, [containerRef, entriesAndIds.length, previousEntriesLength]);

  return (
    <section className="margin-bottom--3" ref={containerRef}>
      {entriesAndIds.map(([entry, id], index) => (
        <RepeatableFieldsetCard
          key={id}
          className="js-repeated-fieldset-card"
          entry={entry}
          heading={`${props.headingPrefix} ${index + 1}`}
          index={index}
          removeButtonLabel={props.removeButtonLabel}
          onRemoveClick={props.onRemoveClick}
          showRemoveButton={entries.length > 1}
        >
          {props.render(entry, index)}
        </RepeatableFieldsetCard>
      ))}

      <Button onClick={props.onAddClick} variation="outline">
        {props.addButtonLabel}
      </Button>
    </section>
  );
};

RepeatableFieldset.propTypes = {
  /**
   * Localized text used in the "Add" button
   */
  addButtonLabel: PropTypes.string.isRequired,
  /** Array of entries, each of which will have the content repeated for. */
  entries: PropTypes.arrayOf(PropTypes.object).isRequired,
  /**
   * Displayed as the heading for each card, followed by the card's position. For example
   * if you specify "Person", headings will be "Person 1", "Person 2", etc.
   */
  headingPrefix: PropTypes.string.isRequired,
  /**
   * Function used for rendering the fields
   * @param {object} entry
   * @param {number} index
   * @returns {React.ReactNode}
   * @see https://reactjs.org/docs/render-props.html
   */
  render: PropTypes.func.isRequired,
  /**
   * Localized text used in the "Remove" buttons.
   * A "Remove" button isn't rendered when the
   * entries length is 1.
   */
  removeButtonLabel: PropTypes.string.isRequired,
  /**
   * Event handler responsible for adding a new entry
   */
  onAddClick: PropTypes.func.isRequired,
  /**
   * Event handler responsible for removing an entry
   * @param {object} entry
   * @param {number} index
   */
  onRemoveClick: PropTypes.func.isRequired,
};

/**
 * Takes an array of entries, and pair each entry with a unique id.
 * The entry-id mapping remains stable across re-renders.
 * Returns an array of [entry, id] pairs.
 * @param {Array} entries List of entries
 * @returns {[*,string][]} List of entry, id pairs
 */
function useEntryIds(entries) {
  const [entryIdMap, setEntryIdMap] = useState(createEntryIdMap(entries));
  useEffect(() => {
    setEntryIdMap(createEntryIdMap(entries, entryIdMap));

    // No need to add entryIdMap to dependency list since entryIdMap
    // only depends on entries and entries is already in the dependency list.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [entries]);
  return Array.from(entryIdMap.entries());
}

/**
 * Create a map from entries to ids.
 * If the entry already existed previously, use the same id.
 * Otherwise, create a new id.
 * @param {Array} entries List of entries
 * @param {Map<*,string>} [prevEntryIdMap] Previous map from entries to ids
 */
function createEntryIdMap(entries, prevEntryIdMap = new Map()) {
  const entryIdMap = new Map();
  for (const entry of entries) {
    const entryId = prevEntryIdMap.get(entry);
    if (entryId) {
      entryIdMap.set(entry, entryId);
    } else {
      entryIdMap.set(entry, uniqueId("RepeatableFieldset"));
    }
  }
  return entryIdMap;
}

export default RepeatableFieldset;
