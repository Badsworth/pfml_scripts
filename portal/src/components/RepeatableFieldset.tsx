import React, { useEffect, useRef, useState } from "react";
import Button from "./Button";
import RepeatableFieldsetCard from "./RepeatableFieldsetCard";
import classnames from "classnames";
import { uniqueId } from "lodash";
import usePreviousValue from "../hooks/usePreviousValue";

interface RepeatableFieldsetProps {
  /**
   * Localized text used in the "Add" button
   */
  addButtonLabel: string;
  /** Array of entries, each of which will have the content repeated for. */
  entries: Array<Record<string, unknown>>;
  /**
   * Displayed as the heading for each card, followed by the card's position. For example
   * if you specify "Person", headings will be "Person 1", "Person 2", etc.
   */
  headingPrefix: string;
  /**
   * Maximum number of repeatable fields
   */
  limit?: number;
  /**
   * Message to display when max number of repeatable fields have been added
   */
  limitMessage?: string;
  /**
   * Function used for rendering the fields
   * @see https://reactjs.org/docs/render-props.html
   */
  render: (entry: Record<string, unknown>, index: number) => React.ReactNode;
  /**
   * Localized text used in the "Remove" buttons.
   * A "Remove" button isn't rendered when the
   * entries length is 1.
   */
  removeButtonLabel: string;
  /**
   * Event handler responsible for adding a new entry
   */
  onAddClick: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Event handler responsible for removing an entry
   */
  onRemoveClick: (entry: Record<string, unknown>, index: number) => void;
}

/**
 * Used for rendering the same set of content and fields for each
 * item in the "entries" prop.
 */
const RepeatableFieldset = (props: RepeatableFieldsetProps) => {
  const { entries } = props;
  const containerRef = useRef<HTMLElement>(null);
  const entriesAndIds = useEntryIds(entries);
  const previousEntriesLength = usePreviousValue(entriesAndIds.length);
  const limitReached = props.limit ? entries.length >= props.limit : false;

  useEffect(() => {
    if (entriesAndIds.length > previousEntriesLength && containerRef.current) {
      // When a new entry is added to the list, focus and scroll it into view.
      const lastEntry = containerRef.current.querySelector(
        ".js-repeated-fieldset-card:last-of-type"
      );
      const focusableElement =
        lastEntry && lastEntry.querySelector("[tabIndex]:first-child, label");

      if (focusableElement instanceof HTMLElement) focusableElement.focus();
    }
  }, [containerRef, entriesAndIds.length, previousEntriesLength]);

  const limitMessageClasses = classnames(
    // full-width on small screens, beside button on larger screens
    "display-block",
    "mobile-lg:display-inline-block",
    // margin between button only on small screens
    "margin-top-1",
    "mobile-lg:margin-top-0",
    // center text to align with spinner on small screens
    "text-center",
    "mobile-lg:text-left"
  );

  return (
    <section className="margin-bottom-4" ref={containerRef}>
      {entriesAndIds.map(([entry, id], index) => (
        <RepeatableFieldsetCard
          key={id}
          className="js-repeated-fieldset-card"
          entry={entry}
          heading={`${props.headingPrefix} ${index + 1}`}
          index={index}
          removeButtonLabel={props.removeButtonLabel}
          onRemoveClick={props.onRemoveClick}
          showRemoveButton={entries.length > 1}
        >
          {props.render(entry, index)}
        </RepeatableFieldsetCard>
      ))}
      <Button
        name="add-entry-button"
        onClick={props.onAddClick}
        variation="outline"
        disabled={limitReached}
      >
        {props.addButtonLabel}
      </Button>
      {limitReached && (
        <strong className={limitMessageClasses}>{props.limitMessage}</strong>
      )}
    </section>
  );
};

/**
 * Takes an array of entries, and pair each entry with a unique id.
 * The entry-id mapping remains stable across re-renders.
 * @returns List of [entry, id] pairs
 */
function useEntryIds(entries: unknown[]) {
  const [entryIdMap, setEntryIdMap] = useState(createEntryIdMap(entries));
  useEffect(() => {
    setEntryIdMap(createEntryIdMap(entries, entryIdMap));

    // No need to add entryIdMap to dependency list since entryIdMap
    // only depends on entries and entries is already in the dependency list.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [entries]);
  return Array.from(entryIdMap.entries());
}

/**
 * Create a map from entries to ids.
 * If the entry already existed previously, use the same id.
 * Otherwise, create a new id.
 */
function createEntryIdMap(
  entries: unknown[],
  prevEntryIdMap: Map<unknown, string> = new Map()
) {
  const entryIdMap = new Map();
  for (const entry of entries) {
    const entryId = prevEntryIdMap.get(entry);
    if (entryId) {
      entryIdMap.set(entry, entryId);
    } else {
      entryIdMap.set(entry, uniqueId("RepeatableFieldset"));
    }
  }
  return entryIdMap;
}

export default RepeatableFieldset;
