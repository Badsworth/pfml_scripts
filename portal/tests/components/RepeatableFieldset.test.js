import { act, render, screen } from "@testing-library/react";
import React from "react";
import RepeatableFieldset from "../../src/components/RepeatableFieldset";
import userEvent from "@testing-library/user-event";

const renderComponent = (customProps = {}) => {
  const props = {
    addButtonLabel: "Add",
    entries: [
      { first_name: "Anton" },
      { first_name: "Bud" },
      { first_name: "Cat" },
    ],
    // eslint-disable-next-line react/display-name
    render: (entry, index) => (
      <p>
        {entry.first_name} â€“ {index}
      </p>
    ),
    headingPrefix: "Person",
    removeButtonLabel: "Remove",
    onAddClick: jest.fn(),
    onRemoveClick: jest.fn(),
  };

  return render(<RepeatableFieldset {...props} {...customProps} />);
};

describe("RepeatableFieldset", () => {
  it("renders a RepeatableFieldsetCard for each entry", () => {
    const { container } = renderComponent();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows an enabled add button without limit message when limit is not reached", () => {
    const limitProps = {
      entries: [{ first_name: "Bud" }],
      limit: 2,
      limitMessage: "You can only add 2",
    };
    renderComponent(limitProps);

    const addButton = screen.getByRole("button");
    expect(addButton).not.toHaveAttribute("disabled");
    expect(screen.queryByText("You can only add 2")).toBeNull();
  });

  it("shows disabled add button and limit message when the limit is reached", () => {
    const limitProps = {
      entries: [{ first_name: "Bud" }, { first_name: "Scooter" }],
      limit: 2,
      limitMessage: "You can only add 2",
    };
    renderComponent(limitProps);

    const addButton = screen.getByRole("button", { name: "Add" });
    expect(addButton).toHaveAttribute("disabled");
    expect(screen.getByText("You can only add 2")).toBeInTheDocument();
  });

  it("does not show a Remove button when only one entry exists", () => {
    renderComponent({ entries: [{ first_name: "Bud" }] });

    expect(screen.queryByRole("button", { name: "Remove" })).toBeNull();
  });

  it("calls onAddClick when the add button is clicked", () => {
    const onAddClickMock = jest.fn();
    renderComponent({ onAddClick: onAddClickMock });
    const addButton = screen.getByRole("button", { name: "Add" });

    userEvent.click(addButton);

    expect(onAddClickMock).toHaveBeenCalledTimes(1);
  });

  it("changes the focused element to the last card's label when a new entry is added", () => {
    const customRender = () => (
      <label className="test-label" htmlFor="field" tabIndex="0">
        Hello
      </label>
    );

    const RepeatableFieldsetWithState = ({ initialEntries }) => {
      const [entries, setEntries] = React.useState(initialEntries);
      const handleAddClick = () => {
        const newEntries = entries.concat([{ first_name: "Dog" }]);
        setEntries(newEntries);
      };
      const props = {
        addButtonLabel: "Add",
        entries,
        // eslint-disable-next-line react/display-name
        render: customRender,
        headingPrefix: "Person",
        removeButtonLabel: "Remove",
        onAddClick: handleAddClick,
        onRemoveClick: jest.fn(),
      };

      return <RepeatableFieldset {...props} />;
    };

    render(
      <RepeatableFieldsetWithState initialEntries={[{ first_name: "Anton" }]} />
    );
    const addButton = screen.getByRole("button", { name: "Add" });
    userEvent.click(addButton);

    const [firstEntry, secondEntry] = screen.getAllByText("Hello");
    expect(firstEntry).not.toHaveFocus();
    expect(secondEntry).toHaveFocus();
  });

  describe.skip("when an entry is removed", () => {
    it("keeps RepeatableFieldsetCard keys stable when an entry is removed", () => {
      const { entries } = props;
      // Use `mount` so that useEffect works
      // see https://github.com/enzymejs/enzyme/issues/2086 and https://github.com/enzymejs/enzyme/issues/2011
      act(() => {
        wrapper = mount(<RepeatableFieldset {...props} />);
      });
      // Need to call wrapper.update since updates were made outside
      // of the act call as part of the useEffect callback
      wrapper.update();

      const keys1 = wrapper
        .find("RepeatableFieldsetCard")
        .map((wrapper) => wrapper.key());

      act(() => {
        // Remove element at index 1
        const newEntries = entries.slice(0, 1).concat(entries.slice(2));
        wrapper.setProps({ entries: newEntries });
      });
      // Need to call wrapper.update since updates were made outside
      // of the act call as part of the useEffect callback
      wrapper.update();

      const keys2 = wrapper
        .find("RepeatableFieldsetCard")
        .map((wrapper) => wrapper.key());

      expect(keys2.length).toBe(keys1.length - 1);
      expect(keys2[0]).toBe(keys1[0]);
      expect(keys2.slice(1)).toEqual(keys1.slice(2));
    });
  });
});
