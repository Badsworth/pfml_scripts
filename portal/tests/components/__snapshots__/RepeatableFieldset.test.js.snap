// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepeatableFieldset renders a RepeatableFieldsetCard for each entry 1`] = `
<section
  className="margin-bottom-4"
>
  <RepeatableFieldsetCard
    className="js-repeated-fieldset-card"
    entry={
      Object {
        "first_name": "Anton",
      }
    }
    heading="Person 1"
    index={0}
    key="RepeatableFieldset1"
    onRemoveClick={[MockFunction]}
    removeButtonLabel="Remove"
    showRemoveButton={true}
  >
    <p>
      Anton
       – 
      0
    </p>
  </RepeatableFieldsetCard>
  <RepeatableFieldsetCard
    className="js-repeated-fieldset-card"
    entry={
      Object {
        "first_name": "Bud",
      }
    }
    heading="Person 2"
    index={1}
    key="RepeatableFieldset2"
    onRemoveClick={[MockFunction]}
    removeButtonLabel="Remove"
    showRemoveButton={true}
  >
    <p>
      Bud
       – 
      1
    </p>
  </RepeatableFieldsetCard>
  <RepeatableFieldsetCard
    className="js-repeated-fieldset-card"
    entry={
      Object {
        "first_name": "Cat",
      }
    }
    heading="Person 3"
    index={2}
    key="RepeatableFieldset3"
    onRemoveClick={[MockFunction]}
    removeButtonLabel="Remove"
    showRemoveButton={true}
  >
    <p>
      Cat
       – 
      2
    </p>
  </RepeatableFieldsetCard>
  <Button
    onClick={[MockFunction]}
    variation="outline"
  >
    Remove
  </Button>
</section>
`;
